File: feishu_h5_doc/batch_convert.py
 # # 指定输入目录，输出默认到 <sourceDir>/output
# python batch_convert.py ./source

# # 指定输入和输出目录
# python batch_convert.py ./source ./dist

#!/usr/bin/env python3
import os
import sys
import mammoth

def sanitize_filename(name):
    """去掉文件名中非法字符"""
    return "".join(c for c in name if c.isalnum() or c in (" ", "_", "-")).rstrip()

def convert_docx_to_html(input_path, output_path, title):
    # Mammoth style map 保留加粗、斜体、下划线
    style_map_text = "\n".join([
        "b => b",
        "strong => b",
        "i => i",
        "em => i",
        "u => u"
    ])
    try:
        with open(input_path, "rb") as docx_file:
            result = mammoth.convert_to_html(
                docx_file,
                style_map=style_map_text,
                ignore_numbering=True   # 忽略列表，避免 KeyError: 'w:ilvl'
            )
            html_body = result.value

        html_template = f"""<!doctype html>
<html lang="zh-CN">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>{title}</title>
<style>
  body {{
    font-family:-apple-system,BlinkMacSystemFont,"Helvetica Neue",Segoe UI,Roboto,"PingFang SC",Arial,sans-serif;
    line-height:1.6;
    margin:16px;
    color:#111827;
  }}
  .table-wrap {{ overflow:auto; }}
  table {{ border-collapse:collapse; width:100%; margin:12px 0; }}
  th, td {{ border:1px solid #ddd; padding:8px; }}
  img {{ max-width:100%; height:auto; }}
</style>
</head>
<body>
<div class="container">
{html_body}
</div>
</body>
</html>"""

        os.makedirs(os.path.dirname(output_path), exist_ok=True)
        with open(output_path, "w", encoding="utf-8") as f:
            f.write(html_template)

        print(f"✅ 转换完成: {input_path} → {output_path}")
    except Exception as e:
        print(f"❌ 转换失败: {input_path} -> {e}")

def batch_convert(source_dir, output_dir):
    os.makedirs(output_dir, exist_ok=True)
    for root, _, files in os.walk(source_dir):
        for file in files:
            if file.endswith(".docx"):
                input_path = os.path.join(root, file)
                base_name = sanitize_filename(os.path.splitext(file)[0])
                rel_path = os.path.relpath(root, source_dir)
                target_dir = os.path.join(output_dir, rel_path)
                os.makedirs(target_dir, exist_ok=True)
                output_path = os.path.join(target_dir, base_name + ".html")
                convert_docx_to_html(input_path, output_path, base_name)

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("用法: python batch_convert.py <sourceDir> [outputDir]")
        sys.exit(1)

    source_dir = sys.argv[1]
    output_dir = sys.argv[2] if len(sys.argv) > 2 else os.path.join(source_dir, "output")

    batch_convert(source_dir, output_dir)

File: feishu_h5_doc/batch-convert.js
// 使用方法：
// node batch-convert.js ./source ./dist
// node batch-convert.js ./source ./dist --docx4js

const fs = require("fs");
const path = require("path");
const mammoth = require("mammoth");
const Docx4js = require("docx4js"); // 3.x

// 获取参数
const [,, sourceDirArg, outputDirArg, extraFlag] = process.argv;
if (!sourceDirArg) {
  console.log("用法: node batch-convert.js <sourceDir> [outputDir] [--docx4js]");
  process.exit(1);
}

const sourceDir = path.resolve(sourceDirArg);
const outputDir = outputDirArg ? path.resolve(outputDirArg) : path.join(sourceDir, "output");
const useDocx4js = extraFlag === "--docx4js";

// 确保输出目录存在
if (!fs.existsSync(outputDir)) {
  fs.mkdirSync(outputDir, { recursive: true });
}

/**
 * Mammoth 转 HTML
 */
async function convertWithMammoth(inputPath, outputPath, title) {
  const options = { styleMap: ["u => u"] }; // 保留下划线
  try {
    console.log(`📄 Mammoth 开始转换: ${inputPath}`);
    const result = await mammoth.convertToHtml({ path: inputPath }, options);
    const htmlBody = result.value;
    const htmlTemplate = wrapHtml(title, htmlBody);
    fs.writeFileSync(outputPath, htmlTemplate, "utf8");
    console.log(`✅ Mammoth 转换完成: ${inputPath} → ${outputPath}`);
  } catch (err) {
    console.error(`❌ Mammoth 转换失败: ${inputPath}`, err);
  }
}

/**
 * Docx4js 3.x 转 HTML
 */
async function convertWithDocx4js(inputPath, outputPath, title) {
  if (!fs.existsSync(inputPath)) {
    console.error(`❌ 文件不存在: ${inputPath}`);
    return;
  }

  console.log(`📄 Docx4js 开始加载文件: ${inputPath}`);
  try {
    // 3.x 版本直接用 Docx4js(path) 或 Docx4js.load(path)
    const doc = await Docx4js.load(inputPath); 
    const htmlBody = doc.getBodyHtml(); // 获取 HTML 内容

    const htmlTemplate = wrapHtml(title, htmlBody);
    fs.writeFileSync(outputPath, htmlTemplate, "utf8");
    console.log(`✅ Docx4js 转换完成: ${inputPath} → ${outputPath}`);
  } catch (err) {
    console.error(`❌ Docx4js 解析失败: ${inputPath}`, err);
  }
}

/**
 * HTML 模板封装
 */
function wrapHtml(title, body) {
  return `<!doctype html>
<html lang="zh-CN">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>${title}</title>
<style>
  :root { --max-width:950px; --padding:16px; --bg:#ffffff; --text:#111827; }
  body { margin:0; padding:0; background:var(--bg); color:var(--text);
    font-family:-apple-system,BlinkMacSystemFont,"Helvetica Neue",Segoe UI,Roboto,"PingFang SC",Arial,sans-serif;
    line-height:1.6; }
  .container { max-width:var(--max-width); margin:24px auto; padding:var(--padding); }
  h1,h2,h3,h4,h5,h6 { margin:18px 0 8px; }
  p { margin:0 0 12px 0; }
  table { width:100%; border-collapse:collapse; margin:12px 0; }
  th, td { border:1px solid #ddd; padding:8px; }
  img { max-width:100%; height:auto; }
  .table-wrap { overflow:auto; }
  @media (max-width:600px) { :root { --padding: 12px; } }
</style>
</head>
<body>
<div class="container">
${body}
</div>
</body>
</html>`;
}

/**
 * 批量转换
 */
async function batchConvert() {
  const files = fs.readdirSync(sourceDir).filter(f => f.endsWith(".docx"));
  for (const file of files) {
    const inputPath = path.join(sourceDir, file);
    const baseName = path.basename(file, ".docx");
    const outputPath = path.join(outputDir, baseName + ".html");

    if (useDocx4js) {
      await convertWithDocx4js(inputPath, outputPath, baseName);
    } else {
      await convertWithMammoth(inputPath, outputPath, baseName);
    }
  }
}

batchConvert();
